torch/csrc/jit/codegen/cuda/ir_nodes.cpp

```diff
@@ -1031,6 +1030,335 @@ std::pair<TensorDomain*, TensorDomain*> TensorDomain::rFactor(
+//! will prove that i2 and i4 are equal in the sense that
+//!    i2.start = i4.start, i2.extent = i4.extent
+//! Depends on ConcretizeDomain, and equalities involving
+//! broadcast domains are defined based on the concretized version
+class ProveValEqual : private IterVisitor {
+ public:
+  explicit ProveValEqual(Fusion* fusion) : cd_(fusion) {
+    traverseFrom(fusion, fusion->outputs(), false);
+  }
+
+  //! Checks if two scalars are equal
+  //!
+  //! First checks if ScalarCheck has them equal,
+  //! next try to prove them equal from
+  //! the graph_traversal result
+  //!
+  //! \param a A symbolic value
+  //! \param b Another value from the same fusion
+  //! \returns Boolean representing if they are proven to be
+  //!          equal based on scalar check and graph traversal
+  bool areEqual(Val* a, Val* b) const {
+    if (ScalarCheck::sameAs(a, b))
+      return true;
+    if (eq_set_.areEquivalent(a, b))
+      return true;
+    return false;
+  }
+
+  //! Checks if two iterdomains are equal
+  //!
+  //! Equality defined as equal start and equal extent
+  //! true means a and b are equal
+  //! false only means that they cannot be proven equal based
+  //! on scalar check and graph traversal
+  //!
+  //! \param a An iterdomain
+  //! \param b Another iterdomain from the same fusion
+  //! \returns Boolean representing if they are proven to be
+  //!          equivalent in the sense that they have equal
+  //!          start and extent
+  bool areEquivalent(IterDomain* a, IterDomain* b) const {
+    if (a->sameAs(b))
+      return true;
+
+    // Abort on un-concretized domains, this can appear once we
+    // allow broadcast on fusion output
+    if (!cd_.canConcretize(a) || !cd_.canConcretize(b))
+      return false;
+
+    auto ac = cd_.concretized(a);
+    auto bc = cd_.concretized(b);
+    return areEqual(ac->start(), bc->start()) &&
+        areEqual(ac->rawExtent(), bc->rawExtent());
+  }
+
+ private:
+  // Utility class to record new equality found
+  void proveId(IterDomain* a, IterDomain* b) {
+    if (!a->sameAs(b)) {
+      eq_set_.join(a->start(), b->start());
+      eq_set_.join(a->rawExtent(), b->rawExtent());
+    }
+  }
+
+  // Inspect a pointwise op and record the identified equality
+  void provePwOp(Expr* e) {
+    TensorView* tv = *ir_utils::filterByType<TensorView>(e->outputs()).begin();
+    std::vector<IterDomain*> io = tv->getRootDomain();
+
+    // Record equalities from output to all the inputs
+    // ignores un-concretizable broadcasts
+    for (auto* i : ir_utils::filterByType<TensorView>(e->inputs())) {
+      std::vector<IterDomain*> ii =
+          TensorDomain::noReductions(i->getMaybeRFactorDomain());
+
+      for (size_t it = 0; it < ii.size(); it++)
+        if (cd_.canConcretize(ii[it]) && cd_.canConcretize(io[it]))
+          proveId(cd_.concretized(ii[it]), cd_.concretized(io[it]));
+    }
+  }
+
+  void handle(ReductionOp* rop) override {
+    provePwOp(rop->asExpr());
```

Comment:
same: `->asExpr()` not needed. Also, when we know the cast is valid, `->as<Expr>()` is prefered (likely we'll deprecate ->asExpr())

Reply:
Updated, thanks!